### ISSUE TITLE
SemanticModel AnalyzeDataFlow returning "wrong" results

### ISSUE THREAD

**Issue by ElemarJR** [dotnet/roslyn#6967]
I’ve been failing to solve this [codecracker issue](https://github.com/code-cracker/code-cracker/issues/562). 

Considering this sample:

``` csharp
using System.Globalization;
using System.Reflection;

namespace ClassLibrary1
{
    public class Class1
    {
        public static void SetDefaultThreadCulture(CultureInfo currentCulture, CultureInfo currentUICulture)
        {
            typeof(CultureInfo).InvokeMember("s_userDefaultCulture", BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.SetField, null, null, new object[] { currentCulture }, CultureInfo.InvariantCulture);
            typeof(CultureInfo).InvokeMember("s_userDefaultUICulture", BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.SetField, null, null, new object[] { currentUICulture }, CultureInfo.InvariantCulture);
            // CultureInfo.DefaultThreadCurrentCulture = new InternationalCultureInfo(); Requres .Net Framework 4.5
        }
    }
}
```

This [analyzer](https://github.com/code-cracker/code-cracker/blob/master/src/CSharp/CodeCracker/Usage/UnusedParametersAnalyzer.cs) is reporting that parameters _currentCulture_ and _currentUICulture_ are not used. But it’s clearly incorrect.

Here is the analyzer code fragment that inspects the code:

``` csharp
private static void Analyzer(SyntaxNodeAnalysisContext context)
{
    if (context.IsGenerated()) return;
    var methodOrConstructor = context.Node as BaseMethodDeclarationSyntax;
    if (methodOrConstructor == null) return;
    var semanticModel = context.SemanticModel;
    if (!IsCandidateForRemoval(methodOrConstructor, semanticModel)) return;
    var parameters = methodOrConstructor.ParameterList.Parameters.ToDictionary(p => p, p => semanticModel.GetDeclaredSymbol(p));
    var ctor = methodOrConstructor as ConstructorDeclarationSyntax;

    if (ctor?.Initializer != null)
    {
        var symbolsTouched = new List<ISymbol>();
        foreach (var arg in ctor.Initializer.ArgumentList.Arguments)
        {
            var dataFlowAnalysis = semanticModel.AnalyzeDataFlow(arg.Expression);
            if (!dataFlowAnalysis.Succeeded) continue;
            symbolsTouched.AddRange(dataFlowAnalysis.ReadInside);
            symbolsTouched.AddRange(dataFlowAnalysis.WrittenInside);
        }
        var parametersToRemove = parameters.Where(p => symbolsTouched.Contains(p.Value)).ToList();
        foreach (var parameter in parametersToRemove)
            parameters.Remove(parameter.Key);
    }

    if (methodOrConstructor.Body.Statements.Any())
    {
        var dataFlowAnalysis = semanticModel.AnalyzeDataFlow(methodOrConstructor.Body);
        if (!dataFlowAnalysis.Succeeded) return;
        foreach (var parameter in parameters)
        {
            var parameterSymbol = parameter.Value;
            if (parameterSymbol == null) continue;
            if (!dataFlowAnalysis.ReadInside.Contains(parameterSymbol) && !dataFlowAnalysis.WrittenInside.Contains(parameterSymbol))
                ReportDiagnostic(context, parameter.Key);
        }
    }
    else
    {
        foreach (var parameter in parameters.Keys)
            ReportDiagnostic(context, parameter);
    }
}
```

 I wrote this unit test to verify the error, but, surprisingly I got “a green”:

``` csharp
[Fact]
public async Task UsedParameterDoesNotCreateDiagnostic2()
{
    const string source = @"
using System.Globalization;
using System.Reflection;

namespace ClassLibrary1
{
public class Class1
{
protected Class1() { }

public static void SetDefaultThreadCulture(CultureInfo currentCulture, CultureInfo currentUICulture)
{
    typeof(CultureInfo).InvokeMember(""s_userDefaultCulture"", BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.SetField, null, null, new object[] { currentCulture }, CultureInfo.InvariantCulture);
    typeof(CultureInfo).InvokeMember(""s_userDefaultUICulture"", BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.SetField, null, null, new object[] { currentUICulture }, CultureInfo.InvariantCulture);
}
}
}
";
    await VerifyCSharpHasNoDiagnosticsAsync(source);
}
```

So, I started to debug and … in runtime (with vs), only WrittenOutside property from resulting DataFlowAnalysis had elements (ReadInside was Empty). When debugging my unit test, ReadInside was correctly reporting two elements.

So, did we some mistake here? Why are Roslyn reporting different results here. Any suggestion how I could “workaround” this?

**Comment by gafter** [comment]
I suspect a different version of Roslyn is being used for the two scenarios: the one built in to the VS you are using, and the one you built a unit test against.

**Comment by jaredpar** [comment]
Closing as haven't heard back on this about whether two Roslyn's were being used.  If we can repro this with a single version please ping us and we'll reopen the issue.
